### PR TITLE
Add the TanStack Start benchmark (Octane vs React) + two compiler fixes it surfaced

### PR DESCRIPTION
**Pull request by trueadm**
## What

`benchmarks/tanstack-start`: **one Start application, built and served twice** — the repo's vendored `@tanstack/octane-start` (native `StreamOptions.injection` stream path, nitro output) versus `@tanstack/react-start` from npm, pinned to the **same release family** as the vendored packages (react-start `1.168.28`, react-router `1.170.18`, minimal srvx server). Both apps derive from TanStack's own MIT-licensed e2e pair at `753f919e`, trimmed to one canonical `start-basic` surface (`/`, `/posts`, `/posts/$postId`, a 404 post, `/deferred`) over a shared deterministic data fixture — no network anywhere.

### Correctness gates (the point of Phase 1 — perf refuses to exist until these are green)

- **`compare.mjs` — structural: 5/5 routes PASS.** Fetches every route from both production servers, strips each framework's dialect (hydration markers, framework scripts/attributes, head ordering, octane's `#__app` container), and requires the remaining element tree + text to match node-for-node (75+ nodes deep on the posts routes).
- **`e2e/bench.spec.ts` — behavioral: 8/8.** One Playwright spec, two projects, identical assertions against both servers with a clean-console gate: hydration, loader/params navigation, streamed deferred resolution, not-found routing.

### Two real octane compiler bugs the build surfaced — fixed with red/green regression tests

1. **Component hoisting**: `function X() @{…}` compiled to a TDZ `const`, so the canonical route-file shape (`createFileRoute(...)({ component: Home })` above the declaration — which TanStack's own octane e2e app uses) crashed at module evaluation, and production swallowed it into empty marker output. Components referenced above their declaration now emit real hoisted function declarations in both compile modes, with capability stamps as `typeof`-guarded statements (guarded because TanStack's code-splitter extracts the declaration and would strand unguarded stamps into `ReferenceError`s — found by this suite's own e2e). Unreferenced components keep the PURE const form, so tree-shaking is unaffected.
2. **Volar template-body markers**: `@tsrx/core` stopped stamping `metadata.native_tsrx_body` on the parse AST, silently no-op'ing the route-generator masker that TanStack's octane flavor uses — route-tree **regeneration** over `.tsrx` crashed babel on `@{`. Invisible until now because every existing app commits `routeTree.gen.ts`. `markNativeTemplateBodies` restores the contract (pinned in `volar.test.ts`), which also protects TanStack/router#7847's future octane upgrade.

### The 8th journey — closed by main

The initially-open octane bug (client child-navigation `/posts → /posts/3` remounting the whole match tree) turned out to be the passthrough-hydration remount defect fixed independently on main (#192): after merging main, all 8 journeys pass and `remount-probe.mjs` confirms byte-identical DOM-identity preservation on both flavors. The committed probes stay as regression tooling. **The perf phase (run.mjs + bench.mjs registration) is now unblocked** and follows as its own PR.

## Testing

- Full suite **11620 passed, zero failures** (both compile modes — including the loader-compatibility regression the first emission draft caused and this suite caught).
- `pnpm typecheck`, `format:check`, `changeset:check` green; evals corpus regenerated for the lockfile change.
- Bench gates as above: `compare` 5/5, e2e 8/8, DOM-identity probe identical across flavors.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Compiler emission changes affect all `.tsrx` route modules and SSR/client parity; benchmark and test coverage mitigate, but hoisting/HMR edge cases could still affect production bundles.
> 
> **Overview**
> Adds **`benchmarks/tanstack-start`**: the same TanStack Start app built twice (vendored **octane-start** + nitro vs pinned **react-start** + srvx), with **`compare.mjs`** (normalized HTML parity on five routes), shared **Playwright** e2e (octane + react projects), **`serve-both.mjs`**, and remount probes. Ignores **`.output`** at the repo root.
> 
> **Octane compiler (patch):** When a component name is referenced **above** its `function … @{…}` declaration (typical `createFileRoute({ component: Home })` before `function Home`), client and server compiles now emit a **hoisted function declaration** instead of a TDZ **`const`**, with **`typeof`-guarded** follow-ups for `__warm`, HMR, and single-root stamps so route code-splitting cannot leave stranded mutations. Names only in strings stay on the PURE **`const`** path.
> 
> **Volar:** **`compileToVolarMappings`** again sets **`metadata.native_tsrx_body`** on native template bodies in **`sourceAst`** for TanStack’s route-generator masker. Regression coverage in **`component-hoisting.test.ts`** and **`volar.test.ts`**; evals manifest **`overlayLockfileHash`** updated.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit b973350a6a6177fbd1392695b92a47efb061a969. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/octane/src/compiler/compile.js [hunk 1]
```diff
@@ -7196,10 +7216,64 @@ function compileComponent(node, ctx, options) {
...
+			statements.push(`${guard}_$__s(${name});`);
+			componentInfo.singleRootInitialized = true;
+		}
+		if (hmrWrap && isExported) {
+			// `_$hmr` returns a stable wrapper with a DIFFERENT identity; rebinding
+			// keeps later references hot-updatable while the pre-declaration capture
+			// deliberately holds the raw function (edits to such a component fall
+			// back to a full reload — the capture predates the wrapper).
+			statements.push(`${guard}(${name} = _$hmr(${name}));`);
+		}
```
[the remark this answers is not in the document]
Comment (trueadm) [reply packages/octane/src/compiler/compile.js:7320]: Checked against the spec and empirically — this one isn't a bug. Export-binding immutability applies on the *importing* side; a module-local binding created by `export function` (or `export default function`) is an ordinary mutable function-declaration binding, and reassigning it inside the module is legal ES. Verified in Node:

```js
export function Named() { return 'raw' }
Named = function wrapper() { return 'wrapped' }
// module evaluates fine; importers see 'wrapped' via the live binding
```

evaluates without error and importers observe the **wrapper** through the live binding — which is exactly the intent here: post-declaration references (and importers) get the hot-updatable `_$hmr` wrapper, while the deliberate carve-out (documented in the emission comment) is that a *pre-declaration* capture holds the raw function and falls back to a full reload on edit. The bench app's vite dev server runs this exact emission (hmr compile + early-referenced exported route components) — that's how the remount investigation's dev probes were driven.

File: benchmarks/tanstack-start/remount-dev-probe.mjs [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import { createRequire } from 'node:module';
+const { chromium } = createRequire(
+	'/Users/trueadm/Projects/octane-tanstack-bench/benchmarks/tanstack-start/package.json',
+)('@playwright/test');
+import { spawn } from 'node:child_process';
+
+const child = spawn('pnpm', ['dev', '--port', '4177'], {
+	cwd: '/Users/trueadm/Projects/octane-tanstack-bench/benchmarks/tanstack-start/octane',
```
[the remark this answers is not in the document]
Comment (trueadm) [reply benchmarks/tanstack-start/remount-dev-probe.mjs]: Fixed in b7792907 — the probe now derives every path from `import.meta.url`, so it runs from any checkout. Verified: builds boot and the probe executes clean from relative paths.

File: benchmarks/tanstack-start/react/src/raw-stream-fns.ts [hunk 3]
```diff
@@ -0,0 +1,522 @@
...
+			};
+		}
+	}
+	return {
+		match: true,
+		mismatchIndex: null,
+		actualLength: actual.length,
+		expectedLength: expected.length,
+	};
+}
```
[the remark this answers is not in the document]
Comment (trueadm) [reply benchmarks/tanstack-start/react/src/raw-stream-fns.ts]: Fixed in b7792907 — `raw-stream-fns.ts`, the `shared-lib` tree, and the custom server-entry smoke file are gone; the react flavor now uses react-start's default server entry. Gates re-verified after the rebuild (compare 5/5, e2e 8/8).

File: benchmarks/tanstack-start/e2e/playwright.config.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+			const address = probe.address();
+			if (address === null || typeof address === 'string') {
+				probe.close();
+				reject(new Error('no port'));
+				return;
+			}
+			probe.close((error) => (error ? reject(error) : resolve(address.port)));
+		});
+	});
+}
```
[the remark this answers is not in the document]
Comment (trueadm) [reply benchmarks/tanstack-start/e2e/playwright.config.ts]: Fixed in b7792907 — `playwright.config.ts` imports `getFreePort` from `serve-both.mjs` instead of reimplementing it.

File: packages/octane/src/compiler/compile.js [hunk 5]
```diff
@@ -7227,10 +7247,64 @@ function compileComponent(node, ctx, options) {
...
```
[the remark this answers is not in the document]
Comment (trueadm) [reply packages/octane/src/compiler/compile.js:7320]: Fixed in 016353e4 — real bug. The inline `_$__s` stamp on the raw function stays (it covers the pre-declaration capture), but the hoisted path no longer claims `singleRootInitialized` when an HMR rebind follows, so the existing module-tail fallback fires and stamps `$$singleRoot` on the wrapper — the binding consumers actually hold. Pinned by an emission-ordering test (tail stamp indexed after the `_$hmr` rebind) plus a prod-mode test asserting the inline stamp remains the only one.

File: packages/octane/src/compiler/compile.js [hunk 6]
```diff
@@ -7227,10 +7247,64 @@ function compileComponent(node, ctx, options) {
...
+	// PURE-initializer form, which bundlers can drop when unused.
+	const sourceBeforeNode =
+		typeof node.start === 'number' && typeof ctx.mapSource === 'string'
+			? // Comments routinely mention component names (docblocks above the
+				// declaration) — strip them so prose can't force the hoisted form.
+				ctx.mapSource.slice(0, node.start).replace(/\/\/[^\n]*|\/\*[\s\S]*?\*\//g, '')
+			: '';
+	const referencedAboveDeclaration =
+		sourceBeforeNode !== '' &&
+		new RegExp(`\\b${name.replace(/\$/g, '\\$')}\\b`).test(sourceBeforeNode);
```
[the remark this answers is not in the document]
Comment (trueadm) [reply packages/octane/src/compiler/compile.js:7283]: Fixed in 016353e4 — the scan now runs through a shared `stripNonReferenceText` helper that blanks comments **and** plain `'…'`/`"…"` string literals before the `\bName\b` test, at both the client and server emission sites, so route paths and import specifiers can no longer force the hoisted form. Template literals are deliberately left intact: `${Name}` interpolations are real evaluated references, and stripping them would reintroduce the TDZ crash this hoisting exists to prevent (documented in the helper). Tests: string-only mention keeps the `/* @__PURE__ */` const form (client + server), interpolation reference still hoists.

File: benchmarks/tanstack-start/remount-probe.mjs [hunk 7]
```diff
@@ -0,0 +1,35 @@
+import { createRequire } from 'node:module';
+const { chromium } = createRequire(process.cwd() + '/package.json')('@playwright/test');
+import { serveBoth } from './serve-both.mjs';
```
[the remark this answers is not in the document]
Comment (trueadm) [reply benchmarks/tanstack-start/remount-probe.mjs:5]: Fixed in b973350a — `createRequire` is now anchored on `import.meta.url` (same as `remount-dev-probe.mjs`), and the probe was re-run from the repo root to confirm it resolves and still reports byte-identical DOM-identity results for both flavors.

File: benchmarks/tanstack-start/react/serve.mjs [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+const server = createServer((req, res) => {
+	const pathname = decodeURIComponent(new URL(req.url, 'http://x').pathname);
+	if (pathname !== '/' && !pathname.includes('..')) {
+		const file = path.join(clientDir, pathname);
+		if (fs.existsSync(file) && fs.statSync(file).isFile()) {
+			res.statusCode = 200;
+			res.setHeader('Content-Type', MIME[path.extname(file)] ?? 'application/octet-stream');
+			fs.createReadStream(file).pipe(res);
+			return;
+		}
```
[the remark this answers is not in the document]
Comment (trueadm) [reply benchmarks/tanstack-start/react/serve.mjs:41]: This one doesn't reproduce — the described behavior is `path.resolve`'s, not `path.join`'s. `join` concatenates segments and never discards the base for an absolute second argument:

```
path.join('/repo/react/dist/client', '/etc/passwd')   // → /repo/react/dist/client/etc/passwd
path.resolve('/repo/react/dist/client', '/etc/passwd') // → /etc/passwd  (what the finding describes)
```

Verified against the running server:
- `GET /assets/deferred-BMYp9JA6.js` → **200** `text/javascript` — client assets ARE resolved and served through this branch (they're what hydrates every e2e journey).
- `GET /etc/passwd` (`--path-as-is`) → **404** via the Start handler's not-found page — the joined path `dist/client/etc/passwd` doesn't exist, so the request falls through.
- `GET /%2e%2e/%2e%2e/%2e%2e/etc/passwd` → **404** — the `..` check runs after `decodeURIComponent`, so encoded traversal is caught too.

Leaving the code as is.
